ES6 test functions in multi-line strings

webbedspace opened the pull request:
This updates the build script, allowing it to interpret ES6 code expressed in function comments instead of concatenated strings (see #190 iscussion) - which is then wrapped in try-catch, eval() and test() structures automatically.

Also, the runtime script has been updated to unwrap those same structures for the tooltip preview, making the resulting display cleaner - lacking even the test() call wrapper.

Only some ES6 and non-standard tests have been converted to make use of this format, for the moment. The remaining tests are the subjects of various pull-requests currently. If those other requests are resolved, conversion of those tests may then be done.

NOTE: as this drastically affects the output of all HTML files, they have not been included in this commit, to allow branch-merging to pass smoothly. A new build commit will thus have to be done after the merge.

webbedspace committed:
ES6 test functions in multi-line strings

This updates the build script, allowing it to interpret ES6 code
expressed in function comments instead of concatenated strings (see #190
discussion) - which is then wrapped in try-catch, eval() and test()
structures automatically.

Also, the runtime script has been updated to unwrap those same
structures for the tooltip preview, making the resulting display cleaner
- lacking even the test() call wrapper.

Only some ES6 and non-standard tests have been converted to make use of
this format, for the moment. The remaining tests are subject to
pull-requests at this moment and will be addressed after those are
resolved.

NOTE: as this drastically affects the output of all HTML files, they
have not been included in this commit, to allow merging to pass
smoothly.

kangax commented on data-es6.js:
```diff
@@ -494,13 +490,9 @@ exports.tests = [
-    }
-  },
+  exec: function() {/*
+    return (function (...args) { return typeof args !== "undefined"; }())
```
Ahh, so refreshing to see this. Feels just like writing ES6 on a spot :)

kangax commented on master.js:
```diff
@@ -56,10 +56,21 @@ domready(function() {
+      var id = "tooltip_" + this.parentNode.id;
+	  
+      infoTooltip.innerHTML = tooltipCache[id] = tooltipCache[id] || scriptEl.innerHTML
+        // unwrap "multi-line comment" code
```
Perhaps we can just save original ES6 code in a data attribute (or hidden tag, etc.) during build, not to bother with deserialization again when showing tooltip? It would be simpler and less fragile.

kangax replied:
We would literally just have to use something like `scriptEl.getAttribute('data-source')` instead of `scriptEl.innerHTML` followed by all the parsing.

webbedspace replied:
OK, done.

webbedspace committed:
Updated/fixed some tests

Converted these tests to multi-line comment strings: super, defaults,
modules, generators, octal and binary literals, template strings. And on
the non-standard side: array comprehensions, for each..in, and generator
comprehensions.

webbedspace committed:
Tooltip preview code now stored as data attr

Instead of converting the innerHTML of the <script> attribute at
runtime, the code is now stuffed onto a HTML attribute on the <script>
element at build time.

Also fixed some whitespace.

webbedspace committed:
Updated some more tests to multi-line strings

Also fixed some capitalisation inconsistencies and updated some old
links.

webbedspace committed:
Updated README.md

kangax commented on README.md:
```diff
@@ -6,3 +7,5 @@ Editing the tests
 
 Edit the `data-es5.js`, `data-es6.js`, `data-es7.js`, or `data-non-standard.js` files to adjust the tests and their recorded browser results. Run `node build.js` to build the HTML files from these JavaScript sources.
+
+The tests themselves should be written in pure ES3, *except* for the sole ES6 feature being tested (as well as any ES5 features strictly required to use the ES6 feature). The test code is placed in multi-line comments (as in [this hack](http://tomasz.janczuk.org/2013/05/multi-line-strings-in-javascript-and.html)), so that node can parse the data scripts without throwing syntax errors. The `build. ...
```
Good idea, thanks!

webbedspace commented:
Just a final note: as I mentioned, this commit didn't update the HTML files because they would create merge conflicts with the other six PRs. So, these changes hasn't actually gone "live" yet. 

You could make a new commit now that updates the HTML files (thus necessitating that I rebase out the resulting conflicts myself), or you could defer that until you've reviewed more PRs.

kangax commented:
I'll take a look at other PR's in the next few days, and I suppose I'll wait with build for now, to avoid conflicts on your side.
